### PR TITLE
Add Events and fix State for SeldonDeployments

### LINKED ISSUE #1560: Surface reconcile errors as Events on the SeldonDeployment
Ensure errors created by the Operator during reconcile are surfaced into the Status part of the custom resource.

### PR DESCRIPTION
**Pull request by cliveseldon**
Fixes #1560 

 * Creates k8s events on core transitions during operator reconcile of a new SeldonDeployment.
 * Ensures State is set to `Failed` and `Description` set if Reconcile fails

Successful deployment

sdep_events_normal

Failed deployment

sdep_events_failed

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **cliveseldon**

If they are not already assigned, you can assign the PR to them by writing `/assign @cliveseldon` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Sat Mar 21 08:38:08 UTC 2020
The logs for [lint] [2] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1570/2.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1570 --build=2

**Comment by seldondev** [comment]
Sat Mar 21 08:38:17 UTC 2020
The logs for [pr-build] [1] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1570/1.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1570 --build=1

**Comment by seldondev** [comment]
Sat Mar 21 09:35:27 UTC 2020
The logs for [integration] [3] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1570/3.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1570 --build=3

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 28d7478521ea3b096287bac0f77cb6e80464e00b | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: operator/config/rbac/role.yaml [hunk 1]
```diff
@@ -6,6 +6,13 @@ metadata:
   creationTimestamp: null
   name: manager-role
 rules:
+- apiGroups:
+  - ""
+  resources:
+  - events
+  verbs:
+  - create
+  - patch
```
Comment (axsaucedo) [inline operator/config/rbac/role.yaml]: Would it not need get as well? Or get is available by default?

File: operator/constants/constants.go [hunk 2]
```diff
@@ -17,3 +17,26 @@ const (
 
 	FirstPortNumber = 9000
 )
+
+const (
+	ControllerName = "seldon-core-controller"
```
Comment (axsaucedo) [inline operator/constants/constants.go]: Was the name not seldon-controller-manager? Or is that for different resource?

Comment (cliveseldon) [reply operator/constants/constants.go]: yes - worth being consistent - will change

File: operator/config/rbac/role.yaml (hunk 1 above)
Comment (cliveseldon) [reply operator/config/rbac/role.yaml]: We are not reading events just creating them, so think its ok

**Comment by seldondev** [comment]
Sat Mar 21 15:47:52 UTC 2020
The logs for [pr-build] [4] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1570/4.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1570 --build=4

**Comment by seldondev** [comment]
Sat Mar 21 15:48:04 UTC 2020
The logs for [lint] [5] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1570/5.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1570 --build=5

**Comment by seldondev** [comment]
Sat Mar 21 15:49:51 UTC 2020
The logs for [lint] [7] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1570/7.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1570 --build=7

**Comment by seldondev** [comment]
Sat Mar 21 15:49:51 UTC 2020
The logs for [pr-build] [6] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1570/6.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1570 --build=6

**Comment by seldondev** [comment]
Sat Mar 21 15:53:16 UTC 2020
The logs for [integration] [8] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1570/8.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1570 --build=8

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | e94daaad6898f39fc4bdb5bd01220ebafc82c319 | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Sat Mar 21 17:08:15 UTC 2020
The logs for [integration] [9] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1570/9.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1570 --build=9

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | e94daaad6898f39fc4bdb5bd01220ebafc82c319 | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Sun Mar 22 08:22:28 UTC 2020
The logs for [integration] [10] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1570/10.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1570 --build=10

**Comment by seldondev** [comment]
Sun Mar 22 09:48:11 UTC 2020
The logs for [lint] [12] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1570/12.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1570 --build=12

**Comment by seldondev** [comment]
Sun Mar 22 09:48:20 UTC 2020
The logs for [pr-build] [11] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1570/11.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1570 --build=11

**Comment by seldondev** [comment]
Sun Mar 22 10:03:57 UTC 2020
The logs for [integration] [13] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1570/13.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1570 --build=13

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | e94daaad6898f39fc4bdb5bd01220ebafc82c319 | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 22881894b8f1c131d64b7f0abd979c9d947cb3de | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Pull request merged**